zerver/webhooks/newrelic/view.py

```diff
@@ -1,49 +1,61 @@
 # Webhooks for external integrations.
-from typing import Any, Dict, Optional
+from typing import Any, Dict
 
 from django.http import HttpRequest, HttpResponse
+from django.utils.translation import ugettext as _
 
 from zerver.decorator import webhook_view
-from zerver.lib.exceptions import UnsupportedWebhookEventType
 from zerver.lib.request import REQ, has_request_variables
-from zerver.lib.response import json_success
-from zerver.lib.validator import check_dict
+from zerver.lib.response import json_error, json_success
 from zerver.lib.webhooks.common import check_send_webhook_message
 from zerver.models import UserProfile
 
-ALERT_TEMPLATE = "{long_description} ([view alert]({alert_url}))."
-
-DEPLOY_TEMPLATE = """
-**{revision}** deployed by **{deployed_by}**:
-
-``` quote
-{description}
-```
+OPEN_TEMPLATE = """
+Incident opened: {condition_name}
+{details}
+[Acknowledge incident]({incident_acknowledge_url})
+[View incident]({incident_url})
+""".strip()
 
-Changelog:
+CLOSE_TEMPLATE = """
+Incident closed: {condition_name}
+{details}
+[View incident]({incident_url})
+""".strip()
 
-``` quote
-{changelog}
-```
+ACKNOWLEDGE_TEMPLATE = """
+Incident acknowledged: {condition_name}
+{details}
+[View incident]({incident_url})
 """.strip()
 
 @webhook_view("NewRelic")
 @has_request_variables
-def api_newrelic_webhook(request: HttpRequest, user_profile: UserProfile,
-                         alert: Optional[Dict[str, Any]]=REQ(validator=check_dict([]), default=None),
-                         deployment: Optional[Dict[str, Any]]=REQ(validator=check_dict([]), default=None),
-                         ) -> HttpResponse:
-    if alert:
-        # Use the message as the subject because it stays the same for
-        # "opened", "acknowledged", and "closed" messages that should be
-        # grouped.
-        subject = alert['message']
-        content = ALERT_TEMPLATE.format(**alert)
-    elif deployment:
-        subject = "{} deploy".format(deployment['application_name'])
-        content = DEPLOY_TEMPLATE.format(**deployment)
+def api_newrelic_webhook(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    payload: Dict[str, Any]=REQ(argument_type='body')
+) -> HttpResponse:
+
+    info = {
+        "condition_name": payload.get('condition_name', 'Unknown condition'),
+        "details": payload.get('details', 'No details.'),
+        "incident_url": payload.get('incident_url', 'https://alerts.newrelic.com'),
+        "incident_acknowledge_url": payload.get('incident_acknowledge_url', 'https://alerts.newrelic.com'),
+    }
+
+    status = payload.get('current_state', None)
+
+    if 'open' in status:
```

Comment:
This will throw exceptions potentially `TypeError: argument of type 'NoneType' is not iterable`

We should check that status isn't None and return json_error early.

Reply:
It's a bit disappointing that mypy isn't detecting this :disappointed:

Reply:
Yeah, that is disappointing. Added a test with that field missing.